### PR TITLE
Pass `NotebookDocument` when invoking `jupyter.notebookeditor.interruptkernel`

### PR DESCRIPTION
**Pull request by joyceerhl**
For https://github.com/microsoft/vscode-jupyter/issues/5242

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [x] Has sufficient logging.
-   [x] Has telemetry for enhancements.
-   [x] Unit tests & system/integration tests are added/updated.
-   [x] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [x] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/datascience/types.ts [hunk 1]
```diff
@@ -590,6 +591,7 @@ export interface INotebookEditor extends Disposable, IInteractiveBase {
      * Once VSC Notebook is stable, this property can be removed.
      */
     readonly type: 'old' | 'custom' | 'native';
+    readonly document: NotebookDocument | undefined;
```
Comment (DonJayamanne) [inline src/client/datascience/types.ts:594]: I'd remove this. 
Basically what this does is we're prolonging the life of `INotebookEditor`.
I'd prefer we don't use such wrappers where possible.
E.g. WebView & Interactive windows don't need this property (yes I can see that from `undefined`.
But we're just adding more stuff to some wrappers and it just keeps getting used elsewheree.

File: src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts [hunk 2]
```diff
@@ -97,10 +99,11 @@ export class NativeEditorCommandListener implements IDataScienceCommandListener
...
 
-    private interruptKernel() {
-        const activeEditor = this.provider.activeEditor;
-        if (activeEditor) {
-            activeEditor.interruptKernel().ignoreErrors();
+    private interruptKernel(document: NotebookDocument | undefined) {
+        const target =
+            this.provider.editors.find((editor) => editor.document === document) ?? this.provider.activeEditor;
+        if (target) {
+            target.interruptKernel().ignoreErrors();
```
Comment (DonJayamanne) [inline src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:106]: Why not just look for an editor with the same Uri.
If. this.provider.activeEditor.uri === document.uri, then we have what we need. else just iterate through the editors & look for the Uri.

This way you don't need to expose the Document in the Editor object.

**Review by DonJayamanne — CHANGES_REQUESTED** [review]
We don't to add `document` property to INotebookditor

File: src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts (hunk 2 above)
Comment (joyceerhl) [reply src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:106]: Yeah that was the other idea I had, but I saw that `document` was already public readonly in VSCNotebook and thought adding it to the interface was cleaner. If we don't want to use `INotebookEditor` I'll make your suggested change.

File: src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts [hunk 3]
```diff
@@ -34,7 +34,9 @@ export class NativeEditorCommandListener implements IDataScienceCommandListener
             commandManager.registerCommand(Commands.NotebookEditorRemoveAllCells, () => this.removeAllCells())
         );
         this.disposableRegistry.push(
-            commandManager.registerCommand(Commands.NotebookEditorInterruptKernel, () => this.interruptKernel())
+            commandManager.registerCommand(Commands.NotebookEditorInterruptKernel, (document: NotebookDocument) =>
```
Comment (rchiodo) [inline src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:37]: When you run this from the command palette doesn't it pass the uri of the file as the first argument? At least execSelectionInInteractive window did that (as in the command palette always passes the URI of the active editor)

Comment (joyceerhl) [reply src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:37]: It doesn't seem to pass anything when invoked from the command palette:

image

Comment (rchiodo) [reply src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:37]: Might depend upon what has focus when the palette comes up?

Comment (IanMatthewHuff) [reply src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:37]: Yeah, I was at least semi-sure that it was the active editor document. It's annoying to track down what parameters are passed for these I've found. I spelunked around in VS Code source before, but I've never found it explicit in the API anywhere that I looked.

Comment (joyceerhl) [reply src/client/datascience/interactive-ipynb/nativeEditorCommandListener.ts:37]: I just tried invoking execSelectionInteractive from the command palette with a selection in a Python file, and I don't see anything being passed as an arg to the handler 😕  I do see the URI being passed when invoked from the context menu, but that's a different scenario. Ian, do you remember how you got that to happen for you?

image


Similarly after setting focus to the notebook cell or the notebook editor, I'm still not seeing the URI getting passed. However at minimum the parameter can clearly be undefined so I should update the type in the handler.

**Review by rchiodo — APPROVED** [review]

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**